backends/ebpf/tests/ptf/common.py

```diff
@@ -208,33 +242,75 @@ def table_write(self, method, table, keys, action=0, data=None, priority=None, r
             cmd = cmd + "ref "
         else:
             cmd = cmd + "id {} ".format(action)
-        cmd = cmd + "key "
-        for k in keys:
-            cmd = cmd + "{} ".format(k)
-        if data:
-            cmd = cmd + "data "
-            for d in data:
-                cmd = cmd + "{} ".format(d)
+        cmd = cmd + self._table_create_str_from_key(keys=keys)
+        cmd = cmd + self._table_create_str_from_data(data=data, counters=counters, meters=meters)
         if priority:
             cmd = cmd + "priority {}".format(priority)
         self.exec_ns_cmd(cmd, "Table {} failed".format(method))
 
-    def table_add(self, table, keys, action=0, data=None, priority=None, references=None):
+    def table_add(self, table, keys, action=0, data=None, priority=None, references=None,
```

Comment:
have you considered adding type annotation to your python? 
I find they help a lot with readability.
So actions are encoded as integers? I wonder whether p4info could be used to use action symbolic names.

Reply:
> have you considered adding type annotation to your python? I find they help a lot with readability.

At some point of time I had an idea to add that annotations, but I abandoned it. I don't remember why. 

> So actions are encoded as integers? I wonder whether p4info could be used to use action symbolic names.

Yes, exactly, actions are encoded as integers. We decided that p4info can't be used at the level of `psabpf[-cli]`, so every think what we can comes from BTF side. DirectCounters and DirectMeters are recognized by name, so for actions it should be also possible, but is not implemented yet.